### PR TITLE
fix: missing user session expiration notification

### PR DESCRIPTION
**Pull request by woyuen**
There was no way to know when a user session has expired.
Add an error event that one can subscribe to handle the expiration properly.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -424,6 +424,17 @@ Returns an object with the following properties:
 
 stelace.auth.logout()
 
+#### User session expiration
+
+To detect when the user session has expired and handle it properly, you can listen to the following error event:
+```
+stelace.onError('userSessionExpired', function () {
+  // remove authentication tokens via the token store
+  // notify users that they must reconnect
```
Comment (abruere) [inline README.md]: Can you add sample code?

```js
stelace.onError('userSessionExpired', function () {
  // remove authentication tokens via the token store, e.g.
  const tokenStore = stelace.getTokenStore()
  tokenStore.removeTokens()
  // In browser, you can also notify users that they must reconnect
  // ...
})
```

File: test/stelace.spec.js [hunk 2]
```diff
@@ -457,3 +458,60 @@ test('Methods return array from list endpoints without pagination', (t) => {
       throw err
     })
 })
+
+test('Emits an event when the user session is expired', (t) => {
```
Comment (abruere) [inline test/stelace.spec.js]: Maybe 'Emits an event when the user session __has__ expired'.

File: test/stelace.spec.js [hunk 3]
```diff
@@ -457,3 +458,60 @@ test('Methods return array from list endpoints without pagination', (t) => {
...
+      clock.tick(3601 * 1000)
+      return stelace.assets.list()
+    })
+    .then(() => {
+      stelace.stopStub()
+      clock.restore()
+      throw new Error('Should not execute this')
+    })
+    .catch(() => {
+      t.is(called, 1)
```
Comment (abruere) [inline test/stelace.spec.js]: May be you should test same request succeeds before ticking the clock (-> 2 calls).

File: lib/stelace.js [hunk 4]
```diff
@@ -179,6 +181,27 @@ export class Stelace {
...
+  onError (type, callback) {
+    if (!errorTypes.includes(type)) throw new Error(`Invalid error type: ${type}`)
+
+    const allListeners = this.getApiField('errorListeners')
+    allListeners[type] = allListeners[type] || []
+    allListeners[type].push(callback)
+
+    return function () {
+      allListeners[type] = allListeners[type].filter(cb => cb !== callback)
+    }
```
Comment (abruere) [inline lib/stelace.js]: You’re not documenting (in readme nor inline with one simple word `return function () { // unsubscribe`) how returned function should be used.

File: README.md (hunk 1 above)
Comment (woyuen) [reply README.md]: Yes, good idea.

File: test/stelace.spec.js (hunk 3 above)
Comment (woyuen) [reply test/stelace.spec.js]: Right, I'll do it.

File: lib/stelace.js (hunk 4 above)
Comment (woyuen) [reply lib/stelace.js]: Thanks, I'll add it in readme.

File: README.md [hunk 5]
```diff
@@ -429,12 +429,24 @@ stelace.auth.logout()
...
-  // ...
+  // Remove authentication tokens via the token store, e.g.
+  const tokenStore = stelace.getTokenStore()
+  tokenStore.removeTokens()
+
+  // In browser, you can also notify users that they must reconnect, e.g.
+  this.notify('Your session has expired.')
+
+  // And removes any information about the current user, e.g.
+  this.$store.dispatch('logout', { sessionExpired: true })
```
Comment (abruere) [inline README.md]: No, this is specific to Vue and can be confusing, please just I include what I suggested.

File: lib/stelace.js (hunk 4 above)
Comment (abruere) [reply lib/stelace.js]: After reading twice, I wonder why we don’t simply name this anonymous function `unsubscribe`, probably better for potential error stacks, and self-documenting.

Comment (abruere) [reply lib/stelace.js]: And you may want to add a test for this too, like having two listeners, unsubscribing one, and checking only the other is called after error.

Comment (woyuen) [reply lib/stelace.js]: True. I added a test to check unsubscribe functions works correctly.

**Comment by woyuen** [comment]
Thanks for the review

**Pull request merged**